### PR TITLE
Avoid querying the chain for outputs for channels we already have

### PR DESCRIPTION
**Pull request by TheBlueMatt**
If we receive a ChannelAnnouncement message but we already have the
channel, there's no reason to do a chain lookup. Instead of
immediately calling the user-provided `chain::Access` when handling
a ChannelAnnouncement, we first check if we have the corresponding
channel in the graph.

Note that if we do have the corresponding channel but it was not
previously checked against the blockchain, we should still check
with the `chain::Access` and update if necessary.

### CODE REVIEW

File: lightning/src/routing/gossip.rs [hunk 1]
```diff
@@ -1433,6 +1433,22 @@ impl<L: Deref> NetworkGraph<L> where L::Target: Logger {
 			return Err(LightningError{err: "Channel announcement node had a channel with itself".to_owned(), action: ErrorAction::IgnoreError});
 		}
 
+		{
+			let channels = self.channels.read().unwrap();
+
+			if let Some(chan) = channels.get(&msg.short_channel_id) {
+				if chan.capacity_sats.is_some() {
```
Comment (dunxen) [inline lightning/src/routing/gossip.rs:1440]: Is `capacity_sats` just one of the fields that would be present if we previously looked it up on-chain? Just wondering if there is something more explicit, but with the comment below I'd say it's pretty clear anyway.

**Review by dunxen — COMMENTED** [review]
Looks good to me besides the (I assume) test fixes

**Review by dunxen — APPROVED** [review]

**Comment by arik-so** [comment]
Are we considering the possibility of reorgs, which might necessitate revalidating the same channel announcement? Not really a big edge case though, so I'm ready to ACK either way.

**Review by arik-so — DISMISSED** [review]

File: lightning/src/routing/gossip.rs [hunk 2]
```diff
@@ -1433,6 +1433,39 @@ impl<L: Deref> NetworkGraph<L> where L::Target: Logger {
...
+					//
+					// Because a reorg could replace one channel with another at the same SCID, if
+					// the channel appears to be different, we re-validate. This doesn't expose us
+					// to any more DoS risk than not, as a peer can always flood us with
+					// randomly-generated SCID values anyway.
+					//
+					// We use the Node IDs rather than the bitcoin_keys to check for "equivalence"
+					// as we didn't (necessarily) store the bitcoin keys, and we only really care
+					// if the peers on the channel changed anyway.
+					if NodeId::from_pubkey(&msg.node_id_1) == chan.node_one && NodeId::from_pubkey(&msg.node_id_2) == chan.node_two {
```
Comment (dunxen) [inline lightning/src/routing/gossip.rs:1452]: At this point, with a reorg, do we care if the capacity might have changed from what we had before for the same SCID?

Comment (dunxen) [reply lightning/src/routing/gossip.rs:1452]: I don't think it's a big deal. We'll definitely figure out how reliable that capacity is anyway.

**Review by dunxen — DISMISSED** [review]

**Review by dunxen — APPROVED** [review]

**Review by arik-so — APPROVED** [review]

**Pull request merged**